src/graph/ChcGraph.cc

```diff
@@ -599,4 +601,126 @@ std::unique_ptr<ChcDirectedHyperGraph> ChcDirectedHyperGraph::makeEmpty(Logic &
+            PTRef var = timeMachine.getVarVersionZero(varName, logic.getSym(vertex)[i]);
+            argVars.insert({pos, var});
+        }
+    }
+
+    vec<PTRef> transitionRelationComponent;
+
+    // KB: creation of the location variables for nodes and argument variables (params of predicates)
+    std::vector<EId> allEdges = getEdges();
+    EdgeTranslator edgeTranslator{*this, locationVars, argVars, {}};
+    // KB: Translation of self-loop edges for the edges vertices
+    //     and edges between vertices in DAG
+    for (auto edge : allEdges) {
+        if (std::find(vertices.begin(), vertices.end(), getSource(edge)) != vertices.end() &&
+            std::find(vertices.begin(), vertices.end(), getTarget(edge)) != vertices.end()) {
+            transitionRelationComponent.push(edgeTranslator.translateEdge(getEdge(edge)));
+            removeEdge(edge);
+        }
+    };
+
+    // KB: Extraction of the state variables of the nested loop nodes, creation of new predicate to replace node with
+    //     nested loops
+    std::vector<PTRef> stateVars = [&locationVars, &argVars]() {
+        std::vector<PTRef> ret;
+        for (auto && entry : locationVars) {
+            ret.push_back(entry.second);
+        }
+        for (auto && entry : argVars) {
+            ret.push_back(entry.second);
+        }
+        return ret;
+    }();
+
+    vec<SRef> args;
+    std::vector<PTRef> vars;
+    for (auto arg : stateVars) {
+        args.push(logic.getSortRef(arg));
+        vars.push_back(timeMachine.getUnversioned(arg));
+    }
+
+    PTRef transitionRelation = logic.mkOr(std::move(transitionRelationComponent));
+    std::string name;
+    for (auto vertex : vertices) {
+        name += logic.getSymName(vertex);
+    }
+
+    SymRef newRef = logic.declareFun(name, logic.getSort_bool(), args);
+    predicates.addRepresentation(newRef, std::move(vars));
+
+    // KB: Changing input and output edges to and from the loop in the graph with new state and loc variables
+    //     Redirecting edges to the newly created vertice
+    allEdges = getEdges();
+    for (auto edge : allEdges) {
+        if (std::find(vertices.begin(), vertices.end(), getSource(edge)) != vertices.end()) {
+            PTRef badStates = locationVars.at(getSource(edge));
+            auto oldEdgeVars = getVariablesFromEdge(logic, *this, edge);
+            std::unordered_map<PTRef, PTRef, PTRefHash> subMap;
+            for (uint i = 0; i < oldEdgeVars.stateVars.size(); i++) {
+                subMap.insert(std::make_pair(oldEdgeVars.stateVars[i], argVars[{getSource(edge), i}]));
+            }
+            updateEdgeSource(edge, newRef);
+            PTRef label = utils.varSubstitute(logic.mkAnd(getEdgeLabel(edge), badStates), subMap);
+
+            updateEdgeLabel(edge, InterpretedFla{label});
+        }
+        if (std::find(vertices.begin(), vertices.end(), getTarget(edge)) != vertices.end()) {
+
+            PTRef initialStates = [&]() -> PTRef {
```

Comment:
`initialStates` is not the right name here.

Reply:
Fixed